tools/base/tests/test_aio.py

```diff
@@ -85,3 +89,1164 @@ async def test_async_subprocess_run(patches, loop, executor):
+    await fired
+
+    async def sad():
+        raise SadError
+
+    def _coro_gen():
+        for coro in coros:
+            if coro == "RAISE":
+                raise LoopError()
+            if coro == "HAPPY":
+                yield happy()
+            elif coro == "SAD":
+                yield sad()
+            elif coro == "FIRED":
+                yield fired
+            else:
+                yield coro
+
+    all_good = all(coro == "HAPPY" for coro in coros)
+    iter_raises = any(coro == "RAISE" for coro in coros)
+
+    if iter_raises:
+        if not iter_type in [iter, aiter]:
+            return
+        _generated_coros = _coro_gen()
+    else:
+        _generated_coros = list(_coro_gen())
+        err_index = next((i for i, x in enumerate(coros) if x != 'HAPPY'), None)
+
+    if iter_type is not None:
+        _coros = iter_type(_generated_coros)
+
+    _results = []
+    _parallel = aio.parallel(_coros, **kwargs)
+
+    if (not all_good and not yield_exceptions) or iter_raises:
+        if iter_raises:
+            with pytest.raises(aio.AsyncParallelIteratorError) as e:
+                async for result in _parallel:
+                    _results.append(result)
+            assert isinstance(e.value.args[0], LoopError)
+            return
+        else:
+            coro_fail = (
+                any(not inspect.isawaitable(coro) for coro in _generated_coros)
+                or any(coro == "FIRED" for coro in coros))
+            if coro_fail:
+                with pytest.raises(aio.AsyncParallelError):
+                    async for result in _parallel:
+                        _results.append(result)
+            else:
+                with pytest.raises(aio.AsyncParallelExecutionError):
+                    async for result in _parallel:
+                        _results.append(result)
+
+        # for iterators there is no way of knowing that more awaitables were
+        # on the way when failure happened, so these need to be closed here
+        if iter_type in (iter, aiter):
+            for coro in _generated_coros[err_index:]:
+                if not isinstance(coro, str):
+                    coro.close()
+
+        if limit < 1 and iter_type != set:
+            # as all jobs are submitted concurrently (the default is higher than
+            # tne number of test jobs, and -1 forces no limit) and as sad is
+            # faster than happy, we get no results
+            assert _results == []
+        elif iter_type != set:
+            # because the ordering on sets is indeterminate the results are unpredictable
+            # therefore the easiest thing is to just exclude them from this test
+            assert _results == coros[:err_index - (err_index % limit)]
+
+        # as we havent closed the loop, and raising an error took priority over cleaning up the
+        #  tasks, gc hasnt happened yet. Any sleep here at all should allow `all_tasks` to return
+        # to its original state.
```

Comment:
might be also worth adding `gc.collect` here as there is the possibility of a very infrequent flake here